File: config/templates/resource_bundle_accessor.m
#import <Foundation/Foundation.h>

@interface BundleFinder_{{MODULE_NAME}} : NSObject
@end

@implementation BundleFinder_{{MODULE_NAME}}
@end

NSBundle* {{MODULE_NAME}}_SWIFTPM_MODULE_BUNDLE() {
  NSString *bundleName = @"{{PACKAGE_NAME}}_{{TARGET_NAME}}";
  NSArray<NSURL *> *candidates = @[
    NSBundle.mainBundle.resourceURL,
    [NSBundle bundleForClass:[BundleFinder_{{MODULE_NAME}} class]].resourceURL,
    NSBundle.mainBundle.bundleURL,
    [NSBundle.mainBundle.bundleURL URLByAppendingPathComponent:@"Frameworks/{{TARGET_NAME}}.framework"]
  ];

  for (NSURL *candidate in candidates) {
    NSURL *bundlePath = [candidate URLByAppendingPathComponent:[bundleName stringByAppendingString:@".bundle"]];
    NSBundle *bundle = [NSBundle bundleWithURL:bundlePath];
    if (bundle) {
      return bundle;
    }
  }
  [NSException raise:NSInternalInconsistencyException format:@"Unable to find bundle named %@", bundleName];
  return nil;
}

File: config/templates/resource_bundle_accessor.swift
import Foundation

private class BundleFinder {}

extension Bundle {
  @available(iOS 8.0, *)
  static let module: Bundle = {
    let bundleName = "{{PACKAGE_NAME}}_{{TARGET_NAME}}"
    let candidates = [
      Bundle.main.resourceURL,
      Bundle(for: BundleFinder.self).resourceURL,
      Bundle.main.bundleURL,
      Bundle.main.bundleURL.appendingPathComponent("Frameworks/{{TARGET_NAME}}.framework")
    ]

    for candidate in candidates {
      let bundlePath = candidate?.appendingPathComponent(bundleName + ".bundle")
      if let bundle = bundlePath.flatMap(Bundle.init(url:)) {
        return bundle
      }
    }
    fatalError("unable to find bundle named \(bundleName)")
  }()
}
